rust/lance/src/dataset/optimize.rs

```diff
@@ -839,6 +696,54 @@ async fn rewrite_files(
     })
 }
 
+async fn rechunk_stable_row_ids(
+    dataset: &Dataset,
+    new_fragments: &mut [Fragment],
+    old_fragments: &[Fragment],
+) -> Result<()> {
+    let mut old_sequences = load_row_id_sequences(dataset, old_fragments)
+        .try_collect::<Vec<_>>()
+        .await?;
+    // Should sort them back into original order.
+    old_sequences.sort_by_key(|(frag_id, _)| {
+        old_fragments
+            .iter()
+            .position(|frag| frag.id as u32 == *frag_id)
+            .expect("Fragment not found")
+    });
+
+    // Need to remove deleted rows
+    futures::stream::iter(old_sequences.iter_mut().zip(old_fragments.iter()))
+        .map(Ok)
+        .try_for_each(|((_, seq), frag)| async move {
+            let deletions = read_deletion_file(&dataset.base, frag, dataset.object_store()).await?;
+            if let Some(deletions) = deletions {
```

Comment:
Are we guaranteed the deletion was materialized at this point?  Is there any way two fragments can be compacted without materializing their deletions?

Reply:
Yeah that is guaranteed. We compacting by scanning and then writing out. The scan removes deleted rows.